### PR TITLE
Adds ami-version-template to manage custom versions.

### PR DESCRIPTION
**Pull request by jajreidy**
Adds AMI version templating as an alternative to the 2 digit versioning. Accepted keywords are major, minor, patch and version. These can be used in any combination. If patch is empty when splitting the version from push_item it will use 0 as a replacement.

Relates to: SPSTRAT-426

### CODE REVIEW

File: src/pubtools/_marketplacesvm/cloud_providers/aws.py [hunk 1]
```diff
@@ -36,6 +36,13 @@ def get_2_digits(version: str) -> str:
         v = version.split(".")[:2]
         return ".".join(v)
 
+    def format_version(version: str, format_template: str) -> str:
+        v = version.split(".")
+        major = v[0]
+        minor = v[1]
+        patch = v[2] or 0
```
Comment (lslebodn) [inline src/pubtools/_marketplacesvm/cloud_providers/aws.py:43]: That could fail if version is just X.Y
```
>>> version = "1.2"
>>> v = version.split(".")
>>> patch = v[2] or 0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list index out of range
```

we might use 
```
major = v.pop(0) if v else None
minor = v.pop(0) if v else None
patch = v.pop(0) if v else None
```

File: src/pubtools/_marketplacesvm/cloud_providers/aws.py [hunk 2]
```diff
@@ -36,6 +36,13 @@ def get_2_digits(version: str) -> str:
...
+        return format_template.format(major=major, minor=minor, patch=patch, version=version)
```
Comment (lslebodn) [inline src/pubtools/_marketplacesvm/cloud_providers/aws.py:44]: Following might for as well. But I am fine with current version. Feel free to resolve
```
return format_template.format(**locals()) 
```

File: src/pubtools/_marketplacesvm/cloud_providers/aws.py [hunk 3]
```diff
@@ -50,7 +57,10 @@ def get_2_digits(version: str) -> str:
     underscore_parts = []
 
     if release.version is not None:
-        underscore_parts.append(get_2_digits(release.version))
+        if ami_version_template:
+            underscore_parts.append(format_version(release.version, ami_version_template))
+        else:
+            underscore_parts.append(get_2_digits(release.version))
```
Comment (lslebodn) [inline src/pubtools/_marketplacesvm/cloud_providers/aws.py:63]: `get_2_digits` should be equivalent of 
```
format_version(release.version, "{major}.{minor}")
```
maybe we could change the default value of the parameter `ami_version_template` to such value :-)

File: src/pubtools/_marketplacesvm/cloud_providers/aws.py (hunk 1 above)
Comment (jajreidy) [reply src/pubtools/_marketplacesvm/cloud_providers/aws.py]: Updated this to be more resilient :)

File: src/pubtools/_marketplacesvm/cloud_providers/aws.py [hunk 4]
```diff
@@ -632,14 +641,8 @@ def _delete_push_images(self, push_item: AmiPushItem, **kwargs) -> Tuple[AmiPush
...
         region = push_item.region or self.default_region
         keep_snapshot = kwargs.get("keep_snapshot") or False
-        name = name_from_push_item(push_item)
         delete_meta_kwargs = {
             "image_id": push_item.image_id,
-            "image_name": name,
-            # currently using getattr for snapshot_id/name because
-            # snapshot related fields are not available in pushitem
-            "snapshot_id": getattr(push_item, "snapshot_id", None),
-            "snapshot_name": getattr(push_item, "snapshot_name", name),
```
Comment (lslebodn) [inline src/pubtools/_marketplacesvm/cloud_providers/aws.py:642]: Is this change for this PR?

File: src/pubtools/_marketplacesvm/cloud_providers/aws.py (hunk 1 above)
Comment (lslebodn) [reply src/pubtools/_marketplacesvm/cloud_providers/aws.py]: TYVM

File: src/pubtools/_marketplacesvm/cloud_providers/aws.py (hunk 2 above)
Comment (lslebodn) [reply src/pubtools/_marketplacesvm/cloud_providers/aws.py]: TYVM

File: src/pubtools/_marketplacesvm/cloud_providers/aws.py (hunk 4 above)
Comment (lslebodn) [reply src/pubtools/_marketplacesvm/cloud_providers/aws.py:642]: OK

**Review by JAVGan — CHANGES_REQUESTED** [review]

File: src/pubtools/_marketplacesvm/cloud_providers/aws.py [hunk 5]
```diff
@@ -32,25 +34,32 @@ def name_from_push_item(push_item: AmiPushItem) -> str:
...
 
     parts.append(release.product)
 
     # Some attributes should be separated by underscores
     underscore_parts = []
 
     if release.version is not None:
-        underscore_parts.append(get_2_digits(release.version))
+        if ami_version_template:
+            underscore_parts.append(format_version(release.version, ami_version_template))
```
[the remark this answers is not in the document]
Comment (jajreidy) [reply src/pubtools/_marketplacesvm/cloud_providers/aws.py:62]: fixed

File: src/pubtools/_marketplacesvm/cloud_providers/aws.py [hunk 6]
```diff
@@ -21,7 +21,9 @@
 UploadResult = namedtuple("UploadResult", "id")  # NOSONAR
 
 
-def name_from_push_item(push_item: AmiPushItem) -> str:
+def name_from_push_item(
```
[the remark this answers is not in the document]
Comment (jajreidy) [reply src/pubtools/_marketplacesvm/cloud_providers/aws.py:24]: fixed

File: src/pubtools/_marketplacesvm/cloud_providers/aws.py [hunk 7]
```diff
@@ -259,7 +268,7 @@ def _get_access_endpoint_url(self, push_item: AmiPushItem) -> Optional[Dict[str,
             }
         return None
 
-    def _copy_image_from_ami_catalog(self, push_item: AmiPushItem, tags: Dict[str, str]):
+    def _copy_image_from_ami_catalog(self, push_item: AmiPushItem, name: str, tags: Dict[str, str]):
```
[the remark this answers is not in the document]
Comment (jajreidy) [reply src/pubtools/_marketplacesvm/cloud_providers/aws.py]: Sounds good to me. applied that

[the remark this answers is not in the document]
Comment (lslebodn) [reply src/pubtools/_marketplacesvm/cloud_providers/aws.py]: IIRC `_copy_image_from_ami_catalog` is used mostly for RHCOS images.
and they do not worry much about AMI name :-)

[the remark this answers is not in the document]
Comment (jajreidy) [reply src/pubtools/_marketplacesvm/cloud_providers/aws.py]: I was making sure it wouldn't break this part:

ami = upload_svc.get_image_by_name(name)

[the remark this answers is not in the document]
Comment (lslebodn) [reply src/pubtools/_marketplacesvm/cloud_providers/aws.py]: sure :-)

**Review by lslebodn — APPROVED** [review]
LGTM

**Pull request merged**